include/fiction/algorithms/network_transformation/node_duplication_planarization.hpp

```diff
@@ -0,0 +1,794 @@
+     */
+    enum class output_order : uint8_t
+    {
+        /**
+         * Keep the PO order from the input network.
+         */
+        KEEP_PO_ORDER,
+        /**
+         * Randomize the PO order.
+         */
+        RANDOM_PO_ORDER
+    };
+    output_order po_order = output_order::KEEP_PO_ORDER;
+};
+
+namespace detail
+{
+
+/**
+ * A structure representing a pair of nodes in an H-graph.
+ *
+ * The nodes stored in this struct describe the fanin-edges of a node in an H-graph.
+ * A node pair object holds two nodes, which are saved in the member 'pair'.
+ * These two outer nodes are connected through zero or more 'middle_nodes'.
+ * The fanin order starts with the first node in 'pair', then proceeds through the 'middle_nodes', and ends with the
+ * second node in 'pair'. The order of 'middle_nodes' is arbitrary as they cannot be further connected to any other
+ * nodes.
+ *
+ * @tparam Ntk Network type for the nodes in the pair.
+ */
+template <typename Ntk>
+struct node_pair
+{
+    /**
+     * Defines the beginning and end of the fanin-edged node.
+     */
+    std::pair<mockturtle::node<Ntk>, mockturtle::node<Ntk>> pair;
+    /**
+     * Specifies the delay value for the node.
+     */
+    uint64_t delay;
+    /**
+     * Contains the nodes between the fanin-edges node; cannot be connected to any other node.
+     */
+    std::vector<mockturtle::node<Ntk>> middle_nodes;
+    /**
+     * Index into the previous level's node_pair vector, used to track fanin-edge alignment.
+     */
+    std::size_t fanin_it{};
+    /**
+     * Standard constructor.
+     *
+     * @param node1 The first node of the fanin-edged node.
+     * @param node2 The second node of the fanin-edged node.
+     * @param delay_value The delay value for the node.
+     */
+    node_pair(mockturtle::node<Ntk> node1, mockturtle::node<Ntk> node2, uint64_t delay_value) :
+            pair(node1, node2),
+            delay(delay_value)
+    {}
+};
+
+/**
+ * Variant of the mockturtle::initialize_copy_network. This function helps with creating new networks from old
+ * networks. In the mockturtle/original version `old2new` is used to map nodes from the old network to nodes in the new
+ * network in a one to one relation. This variant allows old nodes to map to multiple nodes in order to represent
+ * relations to dulicated nodes.
+
+ * A map (old2new) is created where old nodes from source network are mapped to new nodes in destination network.
+ * A destination network is created as a virtual_pi_network<NtkSrc>.
+ *
+ * @tparam Ntk Type of the network.
+ * @param src The source network.
+ *
+ * @return A pair of the destination network and a node map from the source to the destination network.
+ */
+template <typename Ntk>
+std::pair<virtual_pi_network<Ntk>, mockturtle::node_map<std::vector<mockturtle::signal<virtual_pi_network<Ntk>>>, Ntk>>
+initialize_copy_network_duplicates(Ntk const& src)
```

Comment:
Instead of using a `mockturtle::node_map` that maps to an `std::vector` of signals, it is possible to implement your own `node_multimap`. Just think about whether that would make things easier.